src/blueapi/core/context.py

```diff
@@ -112,9 +123,46 @@ def plan_2(...) -> MsgGenerator:
     def with_device_module(self, module: ModuleType) -> None:
         self.with_dodal_module(module)
 
+    def _connect_devices(
+        self, module: ModuleType, devices: dict[str, AnyDevice], **kwargs
+    ):
+        factories = collect_factories(module, include_skipped=False)
+
+        def is_simulated_device(name, factory, **kwargs):
+            device = devices.get(name, None)
+            mock_flag = kwargs.get("mock", kwargs.get("fake_with_ophyd_sim", False))
+            return device is not None and (
+                isinstance(factory, DeviceInitializationController)
+                and (factory._mock or mock_flag)  # noqa: SLF001
+                and isinstance(device, OphydV1Device | OphydV2Device)
+            )
```

Comment:
What we're seeing here is a consequence of a discussion @DiamondJoseph and I had about https://github.com/DiamondLightSource/dodal/pull/881.

The `mock` parameter of the API of `DeviceInitializationController` is meaningless unless `connect_immediately=True`, it is just ignored otherwise and you have to go digging through the factory to find out if it was set. 

You can argue that it should actually be called `mock_if_connect_immediately` or similar.

This is confusing and makes this use case hard so it should be addressed. I can't think of a better way to do this at the moment so we will have to keep it but please make an issue on dodal explaining that you found this hard to implement with the `DeviceInitializationController` API and a corresponding issue in blueapi to simplify this code when we can.

Reply:
I'll mention this in a ticket in dodal